omp/test/distributed/matrix_kernels.cpp

```diff
@@ -0,0 +1,368 @@
+        for (comm_index_type part = 0; part < row_partition->get_num_parts();
+             ++part) {
+            local_d_md_type diag{ref};
+            local_d_md_type offdiag{ref};
+            local_d_md_type d_diag{exec};
+            local_d_md_type d_offdiag{exec};
+            gko::Array<local_index_type> gather_idxs{ref};
+            gko::Array<local_index_type> d_gather_idxs{exec};
+            gko::Array<comm_index_type> recv_offsets{
+                ref, static_cast<gko::size_type>(
+                         row_partition->get_num_parts() + 1)};
+            gko::Array<comm_index_type> d_recv_offsets{
+                exec, static_cast<gko::size_type>(
+                          row_partition->get_num_parts() + 1)};
+            gko::Array<global_index_type> local_to_global_col{ref};
+            gko::Array<global_index_type> d_local_to_global_col{exec};
+
+            gko::kernels::reference::distributed_matrix::build_diag_offdiag(
+                ref, input, row_partition, col_partition, part, diag, offdiag,
+                gather_idxs, recv_offsets.get_data(), local_to_global_col);
+            gko::kernels::omp::distributed_matrix::build_diag_offdiag(
+                exec, d_input, d_row_partition, d_col_partition, part, d_diag,
+                d_offdiag, d_gather_idxs, d_recv_offsets.get_data(),
+                d_local_to_global_col);
+
+            assert_device_matrix_data_equal(diag, d_diag);
+            assert_device_matrix_data_equal(offdiag, d_offdiag);
+            GKO_ASSERT_ARRAY_EQ(gather_idxs, d_gather_idxs);
+            GKO_ASSERT_ARRAY_EQ(recv_offsets, d_recv_offsets);
+            GKO_ASSERT_ARRAY_EQ(local_to_global_col, d_local_to_global_col);
+        }
+    }
+
+    template <typename Data1, typename Data2>
+    void assert_device_matrix_data_equal(Data1& first, Data2& second)
+    {
+        auto size_first = first.get_size();
+        auto size_second = second.get_size();
+        auto arrays_first = first.empty_out();
+        auto array_second = second.empty_out();
+
+        GKO_ASSERT_EQUAL_DIMENSIONS(size_first, size_second);
+        GKO_ASSERT_ARRAY_EQ(arrays_first.row_idxs, array_second.row_idxs);
+        GKO_ASSERT_ARRAY_EQ(arrays_first.col_idxs, array_second.col_idxs);
+        GKO_ASSERT_ARRAY_EQ(arrays_first.values, array_second.values);
+    }
+
+    std::shared_ptr<const gko::ReferenceExecutor> ref;
+    std::shared_ptr<const gko::OmpExecutor> exec;
+    std::default_random_engine engine;
+};
+
+TYPED_TEST_SUITE(Matrix, gko::test::ValueLocalGlobalIndexTypes);
+
+
+TYPED_TEST(Matrix, BuildsDiagOffdiagEmptyIsSameAsRef)
+{
+    using value_type = typename TestFixture::value_type;
+    using local_index_type = typename TestFixture::local_index_type;
+    using global_index_type = typename TestFixture::global_index_type;
+    gko::Array<comm_index_type> mapping{this->ref, {1, 0, 2, 2, 0, 1, 1, 2}};
+    comm_index_type num_parts = 3;
+    auto partition = gko::distributed::Partition<
```

Comment:
nit: Arrange, Act, Assert in three separate blocks
```suggestion

    auto partition = gko::distributed::Partition<
```